timeseries/src/autogluon/timeseries/models/autogluon_tabular/tabular_model.py

```diff
@@ -105,26 +108,132 @@ def _get_features_dataframe(
+
+
+            For example, given the following inputs
+
+            df = [
+                [1, 1, 1, 1],
+                [1, 1, 1, 1],
+                [1, 1, 1, 1],
+            ]
+            num_hidden = [6, 0, 1]
+            lag_indices = [1, 2, 5, 10]
+            num_columns = 1
+
+            The resulting masked dataframe will be
+
+            masked_df = [
+                [NaN, NaN, NaN, 1],
+                [1, 1, 1, 1],
+                [NaN, 1, 1, 1],
+            ]
+
+            """
+            if num_columns > 1:
+                lag_indices = np.repeat(lag_indices, num_columns)
+            mask = num_hidden[:, None] >= lag_indices[None]  # shape [len(num_hidden), len(lag_indices) * num_columns]
+            df[mask] = np.nan
+            return df
+
+        def get_lag_features_and_target(time_series: pd.DataFrame) -> pd.DataFrame:
+            """Construct the dataframe with lagged features and prediction target for a single time series.
+
+            Parameters
+            ----------
+            time_series
+                Dataframe containing a single time series, including target & past/known covariates.
+                Such time series can equivalently be obtained with data.loc[item_id] for some item_id in the dataset.
+
+            Returns
+            -------
+            features
+                Feature dataframe, where each row corresponds to a single timestep of the input time series.
+            """
+            target_lags = get_lags(time_series[[self.target]], self._target_lag_indices)
 
-        def get_lag_features_and_target(group):
-            timestamp = group.index.get_level_values(TIMESTAMP)
-            lag_columns = {f"lag_{idx}": group.shift(idx).values.ravel() for idx in self._lag_indices}
-            features = pd.DataFrame(lag_columns, index=timestamp)
             # Starting from the end of the time series, mask the values as if the last `prediction_length` steps weren't observed
             # This mimics what will happen at test time, when we simultaneously predict the next `prediction_length` values
-            num_windows = (len(group) - 1) // self.prediction_length
+            num_windows = (len(time_series) - 1) // self.prediction_length
             # We don't hide any past values for the first `remainder` values, otherwise the features will be all empty
-            remainder = len(group) - num_windows * self.prediction_length
+            remainder = len(time_series) - num_windows * self.prediction_length
             num_hidden = np.concatenate([np.zeros(remainder), np.tile(np.arange(self.prediction_length), num_windows)])
-            mask = num_hidden[:, None] >= self._lag_indices[None]  # shape [num_timesteps, num_lags]
-            features[mask] = np.nan
+            target_lags = apply_mask(target_lags, num_hidden=num_hidden, lag_indices=self._target_lag_indices)
+            feature_dfs = [target_lags]
+
+            if self.metadata.past_covariates_real:
```

Comment:
wondering if we can support `past_covariates_cat` in the same way here

Reply:
That would also require some other logic throughout the code (e.g., `AbstractTimeSeriesModel.predict_for_scoring`) & require us to infer the covariate type (categorical vs. real). For this reason, I would rather postpone the addition of categorical covariates to v0.8.